### DISCUSSION TITLE
Which one does Jaspr support: "Hot Restart" or "Hot Reload"?

### DISCUSSION THREAD

**Discussion by Hawmex** [schultek/jaspr#163] [General]

**Comment by schultek** [comment]
From  `jaspr serve -h`

```
Runs a development server that serves the web app with SSR and reloads based on file system updates.

Usage: jaspr serve [arguments]
-h, --help                       Print this usage information.
-i, --input                      Specify the input file for the web app.
-m, --mode                       Sets the reload/refresh mode.

          [refresh] (default)    Performs a full page refresh and server reload
          [reload]               Reloads js modules without server reload (loses current state)
```

This relates to the `--auto` option from [`webdev`](https://pub.dev/packages/webdev#usage). Hot-reload is currently not stateful because of this. I haven't had the time to check if it can be made stateful somehow.

Do you have any insights for this? How does dawn work?

**Comment by Hawmex** [comment]
Thanks for the explanation. As I've played around with `webdev`, I've noticed that in the JavaScript environment, unlike the Dart VM, the top-level Dart variables that contain the state are reinitialized during the reload. But that is not the case for variables attached to the `window`.

Example:

```dart
// after each reload, our state is lost.
int count = 0; // assume that it's incremented after some clicks in the app.
```

```dart
import 'package:js/js.dart';


// after each reload, our state is maintained.
@JS()
external int count; // assume that it's later initialized and incremented after some clicks in the app.
```

Maybe we can achieve a stateful hot reload with this approach.

But the next issue is that `runtimeType`'s identity is changed after each hot reload in the JavaScript environment (this is not an issue in Dart VM). Thus, when `runtimeType` is checked in the diffing process, our widget tree gets replaced instead of updated, and therefore `State<T>` is lost.

With all that said, if Dart had behaved the same in the JavaScript environment and the Dart VM, Dawn and Jaspr could support a stateful hot reload by default.

**Comment by Hawmex** [comment]
After all, it is possible to implement a stateful hot reload since Vue has done it.

**Comment by schultek** [comment]
Cool, thanks for this. I will try it out a bit, although the runtimeType issue might become a problem but let's see.
